arch/arm/src/rp2040/rp2040_i2c.c

```diff
@@ -639,7 +642,158 @@ static int rp2040_i2c_transfer(FAR struct i2c_master_s *dev,
 #ifdef CONFIG_I2C_RESET
 static int rp2040_i2c_reset(FAR struct i2c_master_s *dev)
 {
-  return OK;
+  FAR struct rp2040_i2cdev_s *priv = (struct rp2040_i2cdev_s *)dev;
+  unsigned int clock_count;
+  unsigned int stretch_count;
+  uint32_t scl_gpio;
+  uint32_t sda_gpio;
+  uint32_t subsys;
+  uint32_t frequency;
+  int ret;
+
+  DEBUGASSERT(dev);
+
+  /* Our caller must own a ref */
+
+  DEBUGASSERT(priv->refs > 0);
+
+  /* Lock out other clients */
+
+  i2c_takesem(&priv->mutex);
+
+  ret = -EIO;
+
+  /* De-init the port */
+
+  rp2040_i2c_disable(priv);
+
+  /* Use GPIO configuration to un-wedge the bus */
+
+  switch (priv->port)
+    {
+      case 0:
+#if defined(CONFIG_RP2040_I2C0) && CONFIG_RP2040_I2C0_GPIO >= 0
+        sda_gpio = CONFIG_RP2040_I2C0_GPIO;
```

Comment:
Is something like this https://github.com/apache/incubator-nuttx/blob/master/arch/arm/src/stm32h7/stm32_i2c.c#L2596-L2597 possible this the GPIO IP on the RP240.

It would make this have less config option and code.

Reply:
Currently all RP2040 GPIO pins are assigned in https://github.com/apache/incubator-nuttx/blob/master/boards/arm/rp2040/raspberrypi-pico/src/rp2040_boardinitialize.c
The functions, such as I2C and SPI, are assigned here according to the kernel configurations, but the current GPIO API design have no assumption that the assignment may be changed after the boot.  This is the reason of the dirty design in rp2040_i2c_reset().

I think it is better to provide the APIs to retrieve the current GPIO function assignment to make it clean.

Reply:
Ok that makes sense. Yes as it is the GPIO interface seems rather limiting.